_episodes/l2_02_basic_unit_testing.md

```diff
@@ -0,0 +1,293 @@
+
+```bash
+$ g++ your_code.cpp -I path_to_your_gtest.h -L path_to_your_lgtest.a -lgtest_main -lgtest -lpthread -o your_executable_name
+```
+
+
+
+## 4. Test Assertions in GoogleTest
+
+Googletest offers many types of assertions as described in [GoogleTest Assertions](http://google.github.io/googletest/reference/assertions.html). A few of them are described below as they will be used in our upcoming sections/chapters.
+
+<br> </br>
+**1. Equality Assertions**: Equality assertions are used to compare values for equality. The most commonly used assertion is `ASSERT_EQ(expected, actual)`, which verifies that the `expected` and `actual` values are equal. For example:
+
+```cpp
+ASSERT_EQ(expected_value, your_function(function_arguments));  
+// Verify that the expexted value is equal to the value returned by your function.
+
+// You can alo change the order if you prefer.
+ASSERT_EQ(your_function(function_arguments), expected_value);
+```
+
+Other useful equality assertions include `ASSERT_NE`, `ASSERT_LT`, `ASSERT_LE`, `ASSERT_GT`, and `ASSERT_GE` for performing inequality comparisons.
+
+<br> </br>
+**2. Boolean Assertions**: Boolean assertions are used to verify boolean conditions. For example, `ASSERT_TRUE(condition)` checks that the `condition` is `true`, while `ASSERT_FALSE(condition)` ensures that the `condition` is `false`.
+
+```cpp
+ASSERT_TRUE(isValid);  // Verify that the isValid flag is true
+ASSERT_FALSE(hasError);  // Verify that the hasError flag is false
+```
+
+
+<br> </br>
+**3. Exception Assertions**: Exception assertions are used to validate that specific exceptions are thrown during the execution of code. In Google Test, you can use the `ASSERT_THROW(statement, exceptionType)` assertion. For example:
+
+```cpp
+ASSERT_THROW(throwException(), std::runtime_error);  // Verify that throwException() throws a std::runtime_error
+```
+
+
+<br> </br>
+**4. String Assertions**: String assertions are used to compare string values. Google Test provides various string assertions, such as `ASSERT_STREQ`, `ASSERT_STRNE`, `ASSERT_STRCASEEQ`, and `ASSERT_STRCASENE`. These assertions allow you to compare strings for equality, inequality, or case-insensitive equality.
+
+```
+ASSERT_STREQ("Hello", getString());  // Verify that getString() returns the exact string "Hello"
+ASSERT_STRCASEEQ("hello", getLowerCaseString());  // Verify that getLowerCaseString() returns "hello" in a case-insensitive manner
+```
+
+The majority of the macros listed above come as a pair with an `EXPECT_` variant and an `ASSERT_` variant. Upon failure, `EXPECT_` macros generate nonfatal failures and allow the current function to continue running, while `ASSERT_` macros generate fatal failures and abort the current function.
+
+All assertion macros support streaming a custom failure message into them with the `<<` operator, for example:
+
+```cpp
+EXPECT_TRUE(my_condition) << "My condition is not true";
+```
+
+
+<br> </br>
+## 5. Anatomy of a Unit Test
```

Comment:
I've mixed feelings about this going after the example or before the example - there are pros and cons in both cases.

Reply:
As this section delves into the theory in a bit more depth, I think it's ok where it is. We don't want too much theory upfront, I think.

That said, I think we do need a *basic* explanation of what a unit test is before talking about how to work with GoogleTest specifically.